[PXCT-1779] brick docs rework (#2944)
arduino/docs-content: content/software/app-lab/5.bricks/3.custom-bricks/custom-bricks.md

4. Enter a name (ID) for your Brick. This must be a valid Python package name (e.g., lowercase letters, numbers, and underscores).
5. App Lab will generate the folder structure and basic files for your Custom Brick inside the `bricks/` directory.

### Configuration (`brick_config.yaml`)

The `brick_config.yaml` file defines the identity of your Brick.

```yaml
id: my_custom_brick
name: My Custom Brick
description: "A description of what my Custom Brick does."
```

When your App is initialized, the system automatically reads this file and makes the Brick available. Ensure the `id` matches the folder name exactly.

### Python Logic (`__init__.py`)

This is where you write the Python code that interfaces with your App. You expose functionality using the `@brick` decorator provided by the Arduino App Lab utilities.

```python
from arduino.app_utils import brick

@brick
class CustomBrick:
    """My Custom Brick implementation."""

    def __init__(self):
        print("[CustomBrick] Initialized")
    
    def do_something(self):
        print("[CustomBrick] Doing something useful...")
```

### Docker Containers (`brick_compose.yaml`)

> If your Custom Brick requires external services, you can define them using a Docker Compose file. The system will automatically pull and run these containers alongside your App.

Reviewer: ```suggestion
If your Custom Brick requires external services, you can define them in this file. `brick_compose.yaml` is a Docker Compose file:

https://docs.docker.com/reference/compose-file/

The system will automatically pull and run these containers alongside your App.
```

- Adjust wording to make it clear that `brick_compose.yaml` is the referenced Docker Compose file.
- Link to the essential high quality Docker Compose file documentation

Author: 3f3e5a73c9f9b7e15bde362bc156cb20f587a6ad adds the link to `custom-bricks.md`, as well as the new `bricks-reference.md` article.

```yaml
services:
  my_service:

>     image: custom_registry/my_service:latest

Reviewer: ```suggestion
    image: some_registry/some_image:latest
```

- Adjust placeholder reference components to avoid giving the false impression that use of an image created by the user is required. Existing images may also be utilized in custom Bricks.
- Use correct "image" term ([reference](https://docs.docker.com/reference/compose-file/services/#image)).

Author: I've applied the feedback in 466cedd7538e36a8b8d4a452a42e7d77988f57e4

Author: @per1234 In addition to the example with placeholder components, do you think we should include some real examples?

For example, the one you shared [here](https://forum.arduino.cc/t/uno-q-exploring-custom-bricks-and-app-lab-service-containers/1442898/2):

```yaml
services:
  player:
    image: debian:bookworm-slim
    command: sh -c "ls -l /dev/snd || true; sleep infinity"
    devices:
      - /dev/snd
```

Reviewer: >do you think we should include some real examples?

Great idea! I think that would be very helpful to the reader.

> For example, the one you shared [here](https://forum.arduino.cc/t/uno-q-exploring-custom-bricks-and-app-lab-service-containers/1442898/2)

That Compose file was intended only for use directly with Docker for troubleshooting. It doesn't provide an interface that can be used by the Python script of an App.

I suggest providing two demonstrations, one for each fundamental concept:

- Use of a container
- Giving the container access to system resources

## Use of a container

```yaml
services:
  hello_server:
    # Create a container from a Docker image that provides the BSD Netcat networking utility.
    image: toolbelt/netcat
    # Use Netcat to create a web server that responds with a simple message.
    # See: https://manpages.debian.org/unstable/netcat-openbsd/nc.1.en.html
    entrypoint: |
      sh -c ' \
        while true; do
          echo -e "HTTP/1.1 200 OK\r\n\r\nHello, world!" \
          | \
            nc -l -p 5000 -q 0 -v
        done \
      '
    ports:
      # Expose the port for communication with the container's web server.
      # See: https://docs.docker.com/reference/compose-file/services/#ports
      - 5000
```

This Brick provides an interface that can be used from the Python code running on the App's primary container:

```py
import requests

from arduino.app_utils import App

response = requests.get("http://hello_server:5000")
print("Web server response:", response.text)

App.run()
```

Here is a complete demonstration App:

[simple-web-server-brick.zip](https://github.com/user-attachments/files/29425430/simple-web-server-brick.zip)

## Giving the container access to system resources

```yaml
# See: https://docs.docker.com/reference/compose-file/services/
services:
  sound_devices:
    # Create a container from a Docker image that provides the BSD Netcat networking utility.
    image: toolbelt/netcat
    # Make the host system's sound devices available for use inside the container.
    devices:
      - /dev/snd
    # Use Netcat to create a web server that responds with a list of the sound devices present in the container.
    # See: https://manpages.debian.org/unstable/netcat-openbsd/nc.1.en.html
    entrypoint: |
      sh -c ' \
        while true; do
          sound_devices="$$(ls -1 /dev/snd)"

          echo -e "HTTP/1.1 200 OK\r\n\r\n$$sound_devices" \
          | \
            nc -l -p 5000 -q 0 -v
        done \
      '
    ports:
      # Expose the port for communication with the container's web server.
      # See: https://docs.docker.com/reference/compose-file/services/#ports
      - 5000
```

Minimal Python code to interact with the Brick from the App's primary container:

```py
import requests

from arduino.app_utils import App

# Make an HTTP request to the custom Brick's web server.
# See: https://requests.readthedocs.io/en/latest/user/quickstart/#make-a-request
response = requests.get("http://sound_devices:5000")
print("Sound devices:")
print(response.text)

App.run()
```

Demonstration App:

[simple-device-access-brick.zip](https://github.com/user-attachments/files/29425418/simple-device-access-brick.zip)

Author: Thanks Per! These examples are great. My learning preference is definitely to modify and experiment rather than read documentation and implement from scratch. I'm sure many users feel the same way 😄

I added a new ["Custom Brick Examples" section](https://github.com/arduino/docs-content/pull/2944/changes#diff-5e8ecd28164c65936ad71280fc435b17855f23e010012d01e40fd43aa8c3b74eR155-R246).

Author: Related: https://github.com/arduino/docs-content/pull/2944#discussion_r3511794419

    ports:
      - "8080:8080"
```

<Alert type="warning">**Important:** Docker images specified in `brick_compose.yaml` must be publicly accessible, as App Lab does not currently support private registries for Custom Bricks.</Alert>

## Using Your Custom Brick

> Once created, using a Custom Brick is identical to using a built-in Brick. Import the class from your package and initialize it in your App's `main.py`.

Reviewer: ```suggestion
Once created, using a Custom Brick is identical to using a built-in Brick. Import the class from the Python package in the Brick and initialize it in your App's `main.py`.
```

Adjust wording to make it more clear what is being referred by "your package".

Author: See d524f4bee31aae744f6b921cae1c3c0e8ae9229b

```python
from arduino.app_utils import App
from my_custom_brick import CustomBrick

# Initialize the Custom Brick
my_brick = CustomBrick()

# Use its functionality
my_brick.do_something()

# Start the App
App.run()
```

> ## AI Models in Custom Bricks

Reviewer: This is certainly better than no documentation of the feature at all, so I'm fine with shipping the initial version of the docs with this as-is. However, I feel that it doesn't provide nearly enough information for the average user to be able to utilize the feature without reverse engineering how it is done in the built-in Bricks.

Author: I'm working on adding more information, but as long as the current draft is not misleading, let's ship it and tackle the additions in a separate PR.

---
title: Create Custom Bricks in Arduino App Lab
description: Learn how to create and manage Custom Bricks to expand your Apps with custom Python libraries and Docker containers.
overwriteSidebar: Custom Bricks
tags:
  - Bricks
  - Python
  - Docker
  - Custom
  - Arduino App Lab
---

> **Custom Bricks** let you package your own Python modules or third-party Docker services into reusable components. Build a feature once, package it as a Custom Brick, and reuse it across your projects.

Reviewer: ```suggestion
**Custom Bricks** allow you to create additional Docker services and companion Python packages for use by an App.
```

## Reduce prominence of Python component

At least with the current App-local custom Brick feature, the sole value of the Python component of the custom Brick is to facilitate the use of the service/container. If the user only wants to organize their App code into Python modules, they should just add packages/modules under the `python` folder of the App. Using a custom Brick for this purpose only adds unnecessary complexity for absolutely no benefit.

## Replace "module" with "package"

Although it is true that a custom Brick could, provide modules alone, this is not the "golden path". When a custom Brick is created, it is populated as a Python package. Thus, I think it will be quite rare for a custom Brick developer to delete that package and add bare modules instead. For this reason, "Python package" is more likely to be correct than "Python module" in this context.

## Remove mentions of reusability

The App-local custom Brick feature we have currently is mostly just a way to achieve additional functionality for a single App. It is true that the custom Brick could be reused by manually copying it to another App, then manually editing an entry for that brick into the `app.yaml` file. However, doing so would be more of a "hack" than something we support as part of the "golden path".

So it definitely doesn't make sense to advertise reusability as a prominent feature of custom Bricks. Moreover, I feel it would only make sense to mention the possibility of reuse if detailed instructions were provided for the relatively complex task of achieving such a thing.

It probably is best to just wait until such time as proper support for reusable Bricks is implemented (https://github.com/arduino/arduino-app-cli/issues/207).

While Arduino Bricks provide ready-to-use features, Custom Bricks let you integrate any tool or service your application needs.

## Anatomy of a Custom Brick

A Custom Brick is a self-contained package stored locally within your App. Each Custom Brick is located inside the `bricks/` folder within your App's root directory, and the Brick's ID corresponds to the name of its folder. Custom Bricks don't use a namespace prefix (like `arduino:`).

A typical Custom Brick structure looks like this:

```text
my-app/
├── app.yaml
├── python/main.py
└── bricks/
    └── my_custom_brick/            # The Brick ID and folder name

>         ├── __init__.py             # Python logic (required)

Reviewer: ```suggestion
        ├── __init__.py             # Python logic (optional)
```

This file is not required. Here is an example App without a Python package:

https://github.com/user-attachments/files/27430584/Custom.Brick.With.Container.zip

        ├── brick_config.yaml       # Brick metadata and variables (required)
        ├── brick_compose.yaml      # Docker Compose configuration (optional)
        └── requirements.txt        # Python dependencies (optional)
```

<Alert type="info">**Note:** Custom Bricks are local to the App they are created in. They are not shared across different Apps or shown in the official Bricks index.</Alert>

## Create a Custom Brick

You can generate the foundational structure for a Custom Brick directly from the App Lab interface.

1. [Open an App](../../apps/manage-apps/#open-an-app).
2. Click the **Add Brick** button at the top of the **Editor sidebar**.
   ![Screenshot of the App Editor in Arduino App Lab, highlighting the Bricks section in the left sidebar.](../../assets/app-lab-editor-hl-bricks.png)
3. At the bottom of the Bricks catalog, select **Create Custom Brick**.

> 4. Enter a name (ID) for your Brick. This must be a valid Python package name (e.g., lowercase letters, numbers, and underscores). <!-- TODO: Verify -->

Reviewer: ```suggestion
4. Enter a name for your Brick.
```

The name provided by the user is the human-friendly name for the Brick. A compliant ID is automatically generated from that human-friendly name. For example, I created a custom Brick with name "Foo Bar-baz", and this resulted in an ID of `foo_bar_baz`:

```yaml
id: foo_bar_baz
name: Foo Bar-baz
```

The Brick folder is named after the ID, so this results in a valid Python package.

Author: Thanks for confirming!

5. App Lab will generate the folder structure and basic files for your Custom Brick inside the `bricks/` directory.

### Configuration (`brick_config.yaml`)

The `brick_config.yaml` file defines the identity of your Brick and the environment variables it accepts from the App Lab UI.

```yaml
id: my_custom_brick
name: My Custom Brick
description: "A description of what my Custom Brick does."

> category: miscellaneous

Reviewer: ```suggestion
```

These keys don't serve any purpose currently, so there is no reason to include them in this snippet.

variables:
  - name: "MY_SETTING"
    description: "A custom setting for this brick"
    default_value: "123"
```

### Python Logic (`__init__.py`)

Custom Bricks can be implemented in two ways depending on your needs:

#### Simple Function-based Bricks

For utility functions or simple logic, you define plain Python functions. You don't need to use any special decorators.

```python
# bricks/my_custom_brick/__init__.py
def say_hello():
    print("Hello from the custom brick!")
```

#### Managed Class-based Bricks

> If you want your Custom Brick to behave exactly like an official Brick with automatic lifecycle management (background threads, startup/shutdown hooks), you can use the `@brick` decorator.

Reviewer: ```suggestion
If you want your Custom Brick to behave exactly like an official Brick with automatic lifecycle management (background threads, startup/shutdown hooks), use the `@brick` decorator on the class. The App lifecycle management system will automatically call certain methods of the decorated class:
```

- Remove "you can". The optional nature of the use of lifecycle management was already established at the start of the paragraph. If the Brick developer decides they do wish to utilize the lifecycle management feature, the decorator is mandatory, where as "you can" implies it is optional.
- Specify that the decorator is attached to the class.

```python
# bricks/my_custom_brick/__init__.py
from arduino.app_utils import brick
import time

@brick
class MyManagedBrick:
    def start(self):
        print("[Brick] Started")

    @brick.loop
    def my_background_task(self):
        print("[Brick] Running in background...")
        time.sleep(1)
```

> 

Reviewer: ```suggestion

For details on how to utilize the lifecycle management features of the App framework in the Python package of your custom App, see [Bricks Architecture and Configuration Reference](../bricks-reference/).

```

### Docker Containers (`brick_compose.yaml`)

If your Custom Brick requires external services (such as databases or companion APIs), you can define them in this file. The `brick_compose.yaml` file is a standard [Docker Compose file](https://docs.docker.com/reference/compose-file/). The orchestrator will automatically pull and run these containers alongside your App. For full details on Docker capabilities and networking, see [Bricks Architecture and Configuration Reference](../bricks-reference/).

```yaml

>  # bricks/my_custom_brick/brick_compose.yaml

Reviewer: ```suggestion
# bricks/my_custom_brick/brick_compose.yaml
```

Remove stray leading space.

services:
  my_service:
    image: some_registry/some_image:latest
ports:
  - "8080:8080"
```

> <Alert type="warning">**Important:** The orchestrator executes the custom brick's Python code (in `__init__.py`) within the main application's container, **not** inside the custom Docker containers you define here. This means your Python code can't directly access system libraries or files inside `my_database`. Instead, your Python code must communicate with the containerized service over the virtual Docker Compose network using a network API (such as HTTP, WebSockets, or TCP/IP). You can reach the service using the service name defined in your `brick_compose.yaml` (e.g., `my_database`) as the hostname.</Alert>

Reviewer: ```suggestion
<Alert type="warning">**Important:** The orchestrator executes the custom brick's Python code (in `__init__.py`) within the main application's container, **not** inside the custom Docker containers you define here. This means your Python code can't directly access system libraries or files inside the container. Instead, your Python code must communicate with the containerized service over the virtual Docker Compose network using a network API (such as HTTP, WebSockets, or TCP/IP). You can reach the service using the service name defined in your `brick_compose.yaml` (e.g., `my_service`) as the hostname.</Alert>
```

Correct service name.

## Using Your Custom Brick

Once you create a Custom Brick, using it is identical to using a built-in Brick. You import its Python package into your App's `main.py` to instantiate its classes or call its functions.

### Registration (`app.yaml`)

The App Lab UI manages `app.yaml` automatically when you create a Custom Brick.

```yaml
# app.yaml
bricks:
  - my_custom_brick:
      variables:
        MY_SETTING: "123"
```

### Implementation (`python/main.py`)

The orchestrator automatically adds the `bricks/` directory to your Python `sys.path`. You can import the Brick's Python package (its folder name) directly inside `main.py`:

```python
# python/main.py
import os
from arduino.app_utils import App
from my_custom_brick import say_hello, MyManagedBrick

# Call a simple function
say_hello()

# Read the variable passed from the UI
print("Setting:", os.getenv("MY_SETTING"))

# Instantiate a managed brick so App.run() handles it
managed_brick = MyManagedBrick()

# Start the App
App.run()
```

For more details on how to import and initialize Bricks, see [Use Bricks in Your App](../use-bricks/).

## Custom Brick Examples

> These examples illustrate key custom container patterns. Each example includes a complete, downloadable App archive that you can [import into Arduino App Lab](../../apps/manage-apps/#import-an-app).

Reviewer: ```suggestion
These examples illustrate key custom container concepts. Each example includes a complete, downloadable App archive that you can [import into Arduino App Lab](../../apps/manage-apps/#import-an-app).
```

"Pattern" indicates that these could be considered approaches that would be commonly built upon when developing more complex custom Bricks. I'm not sure these could be considered as such. My goal was to produce minimal "Hello, world!" type of demonstrations. It is possible to adopt this approach to some applications, but it is going to be awkward to implement anything complex via the "entrypoint" command alone.

Ideally we would also provide a demonstration of an approach to a custom Brick that is better suited to building on. I would suggest the use of a [Flask](https://flask.palletsprojects.com/)-based web server, something like what is done in this minimal App:

https://github.com/user-attachments/files/27430584/Custom.Brick.With.Container.zip

However, things become significantly more complicated once you move into creating a custom image source, as you explain here:

https://github.com/arduino/docs-content/blob/f66a76c13b0e0ba11788332658638586f4a84eae/content/software/app-lab/5.bricks/4.bricks-reference/bricks-reference.md?plain=1#L136

I think it would be necessary to provide a full tutorial for managing containers built from a custom image source (perhaps something like what is described [here](https://github.com/mjs513/Arduino-Uno-Q-SPI3-Bridge-App/blob/main/README.md#automatic-rebuilds-with-docker-compose-watch) per [this discussion](https://forum.arduino.cc/t/getting-a-neato-xv-11-lidar-working-on-the-q/1445568/31)) along with such a demonstration App.

Author: @per1234 Thanks again! To avoid further expanding the scope of this PR, I'll commit the suggested change only, so we can go ahead with the merge. I've created [PXCT-2036](https://arduino.atlassian.net/browse/PXCT-2036) to track the addition of a dedicated tutorial. You're welcome to comment or reach out directly if you find any other useful examples or resources.

### Use of a container

This example illustrates how to deploy a standard background container and connect to its services over the virtual network.

```yaml
services:
  hello_server:
    # Create a container from a Docker image that provides the BSD Netcat networking utility.
    image: toolbelt/netcat
    # Use Netcat to create a web server that responds with a simple message.
    # See: https://manpages.debian.org/unstable/netcat-openbsd/nc.1.en.html
    entrypoint: |
      sh -c ' \
        while true; do
          echo -e "HTTP/1.1 200 OK\r\n\r\nHello, world!" \
          | \
            nc -l -p 5000 -q 0 -v
        done \
      '
    ports:
      # Expose the port for communication with the container's web server.
      # See: https://docs.docker.com/reference/compose-file/services/#ports
      - 5000
```

> This Brick provides an interface that can be used from the Python code running on the App's primary container:

Author: @per1234 [Kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/) seems to use "primary container" and "main container" interchangeably. Given the `main.py` filename, I think that "main container" would be a more consistent terminology.

```suggestion
This Brick provides an interface that can be used from the Python code running on the App's main container:
```

Reviewer: "Main container" is fine with me.

### Giving the container access to system resources

This example shows how to configure your Docker service to access physical devices exposed to the container.

```yaml
# See: https://docs.docker.com/reference/compose-file/services/
services:
  sound_devices:
    # Create a container from a Docker image that provides the BSD Netcat networking utility.
    image: toolbelt/netcat
    # Make the host system's sound devices available for use inside the container.
    devices:
      - /dev/snd
    # Use Netcat to create a web server that responds with a list of the sound devices present in the container.
    # See: https://manpages.debian.org/unstable/netcat-openbsd/nc.1.en.html
    entrypoint: |
      sh -c ' \
        while true; do
          sound_devices="$$(ls -1 /dev/snd)"

          echo -e "HTTP/1.1 200 OK\r\n\r\n$$sound_devices" \
          | \
            nc -l -p 5000 -q 0 -v
        done \
      '
    ports:
      # Expose the port for communication with the container's web server.
      # See: https://docs.docker.com/reference/compose-file/services/#ports
      - 5000
```

> Minimal Python code to interact with the Brick from the App's primary container:

Author: ```suggestion
Minimal Python code to interact with the Brick from the App's main container:
```

@per1234 Same as for https://github.com/arduino/docs-content/pull/2944/changes#r3493608128.
